velox/connectors/hive/iceberg/IcebergPartitionIdGenerator.cpp

```diff
@@ -0,0 +1,209 @@
+ * limitations under the License.
+ */
+
+#include "velox/connectors/hive/iceberg/IcebergPartitionIdGenerator.h"
+
+#include "velox/connectors/hive/HivePartitionUtil.h"
+#include "velox/connectors/hive/iceberg/IcebergPartitionPath.h"
+#include "velox/connectors/hive/iceberg/TransformEvaluator.h"
+#include "velox/functions/prestosql/URLFunctions.h"
+
+namespace facebook::velox::connector::hive::iceberg {
+
+namespace {
+
+template <TypeKind Kind>
+std::pair<std::string, std::string> makePartitionKeyValueString(
+    const IcebergPartitionUtilPtr& formatter,
+    const BaseVector* partitionVector,
+    vector_size_t row,
+    const std::string& name,
+    const TypePtr& type) {
+  using T = typename TypeTraits<Kind>::NativeType;
+  if (partitionVector->isNullAt(row)) {
+    return std::make_pair(name, "null");
+  }
+
+  return std::make_pair(
+      name,
+      formatter->toPartitionString(
+          partitionVector->as<SimpleVector<T>>()->valueAt(row), type));
+}
+
+} // namespace
+
+IcebergPartitionIdGenerator::IcebergPartitionIdGenerator(
+    const RowTypePtr& inputType,
+    std::vector<column_index_t> partitionChannels,
+    IcebergPartitionSpecPtr partitionSpec,
+    uint32_t maxPartitions,
+    const ConnectorQueryCtx* connectorQueryCtx)
+    : PartitionIdGenerator(
+          partitionChannels,
+          maxPartitions,
+          connectorQueryCtx->memoryPool(),
+          /*partitionPathAsLowerCase=*/false),
+      partitionSpec_(partitionSpec),
+      connectorQueryCtx_(connectorQueryCtx),
+      transformEvaluator_(
+          std::make_unique<TransformEvaluator>(
+              TransformExprBuilder::toExpressions(
+                  partitionSpec_,
+                  partitionChannels_,
+                  inputType),
+              connectorQueryCtx_)) {
+  // Build partition key types and names from transforms.
+  // For each transform, create a hasher for its result type and build the
+  // partition key name (either source column name for identity transforms,
+  // or "columnName_transformName" for other transforms).
+  std::vector<TypePtr> partitionKeyTypes;
+  std::vector<std::string> partitionKeyNames;
+  column_index_t idx{0};
+  for (const auto& field : partitionSpec_->fields) {
+    hashers_.emplace_back(
+        exec::VectorHasher::create(field.resultType(), idx++));
+
+    VELOX_USER_CHECK(
+        hashers_.back()->typeSupportsValueIds(),
+        "Unsupported partition type: {}.",
+        field.resultType()->toString());
+
+    partitionKeyTypes.emplace_back(field.resultType());
+    std::string key = field.transformType == TransformType::kIdentity
+        ? field.name
+        : fmt::format(
+              "{}_{}",
+              field.name,
+              TransformTypeName::toName(field.transformType));
+    partitionKeyNames.emplace_back(std::move(key));
+  }
+
+  rowType_ = ROW(std::move(partitionKeyNames), std::move(partitionKeyTypes));
+
+  partitionValues_ = BaseVector::create<RowVector>(
```

Comment:
We have a getter that returns partitionValues_ as is. It seems this vector may have more rows than are actually valid / populated. Is this intentional?